gapic/src/main/com/google/gapid/util/StatusWatcher.java

```diff
@@ -204,4 +205,146 @@ public String getStatusLabel() {
+      Replay replay = replays.get(update.getReplayId());
+      if (replay == null) {
+        if (update.getStatus() == Service.ReplayStatus.REPLAY_FINISHED) {
+          // Got a replay finished message for a replay we've not seen before. Just ignore it.
+          return false;
+        }
+        replay = new Replay(update);
+        replays.put(update.getReplayId(), replay);
+
+        switch (replay.status) {
+          case REPLAY_QUEUED:
+            queued++;
+            return true;
+          case REPLAY_STARTED:
+            started++;
+            return true;
+          case REPLAY_EXECUTING:
+            executing++;
+            doneInstr += replay.doneInstr;
+            totalInstr += replay.totalInstr;
+            return true;
+          default:
+            return false;
+        }
+      }
+
+      switch (replay.status) {
+        case REPLAY_QUEUED:
+          switch (update.getStatus()) {
+            case REPLAY_STARTED:
+              queued--;
+              return started(replay);
+            case REPLAY_EXECUTING:
+              queued--;
+              return executing(replay, update);
+            case REPLAY_FINISHED:
+              queued--;
+              return finished(replay);
+            default:
+              return false;
+          }
+        case REPLAY_STARTED:
+          switch (update.getStatus()) {
+            case REPLAY_EXECUTING:
+              started--;
+              return executing(replay, update);
+            case REPLAY_FINISHED:
+              started--;
+              return finished(replay);
+            default:
+              return false;
+          }
+        case REPLAY_EXECUTING:
+          switch (update.getStatus()) {
+            case REPLAY_EXECUTING:
+              return executing(replay, update);
+            case REPLAY_FINISHED:
+              executing--;
+              return finished(replay);
+            default:
+              return false;
+          }
+        default:
+          return false;
+      }
+    }
+
+    private boolean started(Replay replay) {
+      started++;
+      replay.status = Service.ReplayStatus.REPLAY_STARTED;
+      return true;
+    }
+
+    private boolean executing(Replay replay, Service.ReplayUpdate update) {
+      int done = update.getFinishedInstrs();
+      int total = update.getTotalInstrs();
+      if (replay.status != Service.ReplayStatus.REPLAY_EXECUTING) {
+        executing++;
+        doneInstr += done;
+        totalInstr += total;
+      } else {
+        doneInstr += done - replay.doneInstr;
+        totalInstr += total - replay.totalInstr;
+      }
+      replay.status = Service.ReplayStatus.REPLAY_EXECUTING;
+      replay.doneInstr = done;
+      replay.totalInstr = total;
+      return true;
+    }
+
+    private static boolean finished(Replay replay) {
+      // TODO: we should clean these out after some time.
+      replay.status = Service.ReplayStatus.REPLAY_FINISHED;
+      return true;
```

Comment:
I think we should do       
```doneInstr -= replay.doneInstr; totalInstr -= replay.totalInstr;```
here, otherwise the doneInstr and totalInstr will be non-decreasing, and after a few replays the status progress will start from a large number like 95% (1+100000/100+100000)

Reply:
Done. Thanks!